collect_app/src/main/java/org/odk/collect/android/utilities/FormDownloader.java

```diff
@@ -0,0 +1,726 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android.utilities;
+
+import android.content.ContentValues;
+import android.database.Cursor;
+import android.net.Uri;
+
+import org.javarosa.xform.parse.XFormParser;
+import org.kxml2.kdom.Element;
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.dao.FormsDao;
+import org.odk.collect.android.listeners.FormDownloaderListener;
+import org.odk.collect.android.logic.FormDetails;
+import org.odk.collect.android.logic.MediaFile;
+import org.odk.collect.android.provider.FormsProviderAPI;
+import org.opendatakit.httpclientandroidlib.Header;
+import org.opendatakit.httpclientandroidlib.HttpEntity;
+import org.opendatakit.httpclientandroidlib.HttpResponse;
+import org.opendatakit.httpclientandroidlib.HttpStatus;
+import org.opendatakit.httpclientandroidlib.client.HttpClient;
+import org.opendatakit.httpclientandroidlib.client.methods.HttpGet;
+import org.opendatakit.httpclientandroidlib.protocol.HttpContext;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+
+import timber.log.Timber;
+
+public class FormDownloader {
+
+    private static final String MD5_COLON_PREFIX = "md5:";
+    private static final String TEMP_DOWNLOAD_EXTENSION = ".tempDownload";
+
+    private FormDownloaderListener stateListener;
+
+    private FormsDao formsDao;
+
+    public void setDownloaderListener(FormDownloaderListener sl) {
+        synchronized (this) {
+            stateListener = sl;
+        }
+    }
+
+    private static final String NAMESPACE_OPENROSA_ORG_XFORMS_XFORMS_MANIFEST =
+            "http://openrosa.org/xforms/xformsManifest";
+
+    static boolean isXformsManifestNamespacedElement(Element e) {
+        return e.getNamespace().equalsIgnoreCase(NAMESPACE_OPENROSA_ORG_XFORMS_XFORMS_MANIFEST);
+    }
+
+    private class TaskCancelledException extends Exception {
+        private final File file;
+
+        TaskCancelledException(File file) {
+            super("Task was cancelled during processing of " + file);
+            this.file = file;
+        }
+
+        TaskCancelledException() {
+            super("Task was cancelled");
+            this.file = null;
+        }
+    }
+
+    public HashMap<FormDetails, String> downloadForms(List<FormDetails> toDownload) {
```

Comment:
`Map<FormDetails, String>` would be a better return type, it hides the underlying Map implementation.

Reply:
Yes but it's the code I just moved from one class to another. Since this pr is pretty big itself we should do such a refactoring in a separate one.

Reply:
Perhaps we need another opinion as to whether the comments are worth while pursuing and then capture them elsewhere?

Reply:
I think the cleanup suggested here and in other comments is worthwhile. @dcbriccetti tends to do this kind of cleanup using save Android Studio refactor commands and commit them first before starting modifications on a piece of code and that's what I'd tend to do now as well. But doing it as a separate PR is also totally fine.

Reply:
Hi @lognaturel, I'm just learning the process a bit here.. to raise as another PR, is it preferable to raise an issue first to capture the suggestions? (Similar to https://github.com/opendatakit/collect/issues/943)

Reply:
In general, yes. For smaller code improvements like the ones we're talking about here I'm really not too picky and would rather reduce friction so they actually get done. As long as there's some context in the PR description and some description of why the changes are safe, I think it's fine not to have an issue. Others may disagree!